log4j-api/src/main/java/org/apache/logging/log4j/spi/Provider.java

```diff
@@ -326,12 +326,10 @@ public String getThreadContextMap() {
             return threadContextMapClass;
         }
         // Default based on properties
-        if (Constants.ENABLE_THREADLOCALS) {
-            return props.getBooleanProperty(GC_FREE_THREAD_CONTEXT_PROPERTY)
-                    ? GC_FREE_THREAD_CONTEXT_PROPERTY
-                    : COPY_ON_WRITE_CONTEXT_MAP;
+        if (props.getBooleanProperty(GC_FREE_THREAD_CONTEXT_PROPERTY, !Constants.IS_WEB_APP)) {
+            return GC_FREE_THREAD_CONTEXT_PROPERTY;
         }
-        return WEB_APP_CONTEXT_MAP;
+        return Constants.IS_WEB_APP ? WEB_APP_CONTEXT_MAP : COPY_ON_WRITE_CONTEXT_MAP;
```

Comment:
This change probably merits a longer discussion.

If the user does not provide a value for `enableThreadlocals`, the type of thread context map is determined by `isWebapp` and `garbagefreeThreadContextMap` in the following way:

## Before

|          |       | garbagefreeThreadContextMap |             |             |
|----------|-------|-----------------------------|-------------|-------------|
|          |       | `false`                       | `true`        | `null`        |
| **isWebapp** | `false` | CopyOnWrite                 | GargageFree | CopyOnWrite |
|          | `true`  | WebApp                      | WebApp      | WebApp      |


## After

|          |       | garbagefreeThreadContextMap |             |             |
|----------|-------|-----------------------------|-------------|-------------|
|          |       | `false`                       | `true`        | `null`        |
| **isWebapp** | `false` | CopyOnWrite                 | GarbageFree | GarbageFree |
|          | `true`  | WebApp                      | GarbageFree      | WebApp      |